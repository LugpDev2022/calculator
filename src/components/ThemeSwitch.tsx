import sun from '/sun.png';
import moon from '/moon.png';

interface Props {
  isThemeDark: boolean;
  handleChange: () => void;
}

const ThemeSwitch: React.FC<Props> = ({ isThemeDark, handleChange }) => {
  const switchContainerStyle = `
    ${isThemeDark ? 'bg-[#0A0B0D]' : 'bg-white'}
    cursor-pointer
    relative
    rounded-full
    transition
    h-8
    w-[72px]
  `;

  const switchInputStyle = 'sr-only peer';

  const sunIconStyle = `
    absolute
    h-[calc(100%-4px)]
    top-0.5
    left-0.5
  `;

  const moonIconStyle = `
    absolute
    h-[calc(100%-4px)]
    top-0.5
    right-[3.5px]
  `;

  const switchSliderStyle = `
    absolute
    aspect-square
    bg-slate-400
    peer-checked:bg-amber-400
    duration-500
    h-[calc(100%-4px)]
    left-0.5
    peer-checked:left-[40.5px]
    rounded-full
    top-0.5
    transition-all
  `;

  return (
    <label className={switchContainerStyle}>
      <input
        type='checkbox'
        className={switchInputStyle}
        onChange={handleChange}
        checked={!isThemeDark}
      />
      <img src={sun} className={sunIconStyle} alt='Sun Icon' />
      <img src={moon} className={moonIconStyle} alt='Moon Icon' />
      <span className={switchSliderStyle}></span>
    </label>
  );
};

export default ThemeSwitch;
