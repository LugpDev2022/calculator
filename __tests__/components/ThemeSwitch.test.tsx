import React from 'react';
import { vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import ThemeSwitch from '../../src/components/ThemeSwitch';

describe('Tests on <ThemeSwitch />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('should match snapshot', () => {
    const { container } = render(<ThemeSwitch handleChange={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  test('should call handleChange on click', () => {
    const handleChange = vi.fn();
    render(<ThemeSwitch handleChange={handleChange} />);
    const themeSwitch: any = document.getElementById('Switch-Container');
    fireEvent.click(themeSwitch);

    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  test('should display the light theme', () => {
    render(<ThemeSwitch handleChange={() => {}} />);

    const checkbox: HTMLInputElement = screen.getByRole('checkbox');
    expect(checkbox.checked).toBeTruthy();
  });

  test('should display the dark theme', () => {
    render(<ThemeSwitch handleChange={() => {}} isThemeDark />);

    const checkbox: HTMLInputElement = screen.getByRole('checkbox');
    expect(checkbox.checked).toBeFalsy();
  });
});
