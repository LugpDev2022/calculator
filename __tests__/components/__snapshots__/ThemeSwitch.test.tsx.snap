// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on <ThemeSwitch /> > should match snapshot 1`] = `
<div>
  <label
    class=" switch-container cursor-pointer relative rounded-full transition h-8 w-[72px] "
    data-testid="switch-container"
  >
    <input
      checked=""
      class="sr-only peer"
      type="checkbox"
    />
    <img
      alt="Sun Icon"
      class="
    absolute
    h-[calc(100%-4px)]
    top-0.5
   left-0.5"
      src="/sun.png"
    />
    <img
      alt="Moon Icon"
      class="
    absolute
    h-[calc(100%-4px)]
    top-0.5
   right-[3.5px]"
      src="/moon.png"
    />
    <span
      class=" absolute aspect-square bg-slate-400 peer-checked:bg-amber-400 duration-500 h-[calc(100%-4px)] left-0.5 peer-checked:left-[40.5px] rounded-full top-0.5 transition-all "
    />
  </label>
</div>
`;
